File: C++/Minesweeper/board.cpp
#include <iostream>
#include <stdlib.h>
#include <time.h>

#include "board.h"

using namespace std;

void Board::Initialize() {
	// No check on input should be added
	cout << "Height: ";
	cin >> height;
	cout << "Width: ";
	cin >> width;
	cout << "Number of mines: ";
	cin >> mineCount;

	board = new int*[height];
	mask = new char*[height];

	for(int i = 0; i < height; i++){
		board[i] = new int[width];
		mask[i] = new char[width];
	}

	for(int i = 0; i < height; i++){
		for(int j = 0; j < width; j++){
			board[i][j] = 0;
			mask[i][j] = '#';
		}
	}
}

void Board::Play() {
	Initialize();
	PlaceMines();
	PrintMask();
	cout << endl;

	int h, w;

	bool lost = false;
	while (!lost) {
		cin >> h >> w;
		lost = OpenField(h, w);
		PrintMask();
	}
	cout << "Game over" << endl;
	RevealBoard();
	PrintMask();
}

void Board::PlaceMines() {
	srand(time(NULL));

	int minesPlaced = 0;

	while (minesPlaced < mineCount) {
		int randH = rand()%(height);
		int randW = rand()%(width);

		if (board[randH][randW] == 9) {
			break;
		}
		else {
			board[randH][randW] = 9;
			minesPlaced++;
		}

		int lowY = randH - 1; 
		int lowX = randW - 1;

		int highY = randH + 1;
		int highX = randW + 1;

		if(randH == 0){
			lowY = 0;
		}
		if(randW == 0){
			lowX = 0;
		}
		if(randH == height){
			lowY = height;
		}
		if(randW == width){
			lowY = width;
		}

		for(int i = lowY; i < highY + 1; i++){
			for(int j = lowX; j < highX + 1; j++){
				if(i == height){
					break;
				}
				if(j == width){
					break;
				}
				if(board[i][j] == 9){
					board[i][j] == 9;
				}
				else{
					board[i][j]++;
				}
			}
		}
	}
}

bool Board::OpenField(int h, int w) {
	// Returns true if player hit a mine
	if (board[h][w] == 9) {
		return true;
	}
	else {
		char i = board[h][w];
		mask[h][w] = '0' + i;
	}
	return false;
}

void Board::RevealBoard() {
	for (int i = 0; i < height; ++i) {
		for (int j = 0; j < width; ++j) {
			if (board[i][j] == 9) {
				mask[i][j] = 'x';
			}
			else {
				mask[i][j] = '0' + board[i][j];
			}
		}
	}
}

void Board::CleanUp() {
	for (int i = 0; i < height; ++i){
		delete [] board[i];
		delete [] mask[i];
	}
	delete [] board;
	delete [] mask;
}

void Board::PrintMask() {
	cout << "    ";
	for (int i = 0; i < width; ++i) {
		cout << i%10 << " ";
	}
	cout << endl << "  ";
	for (int i = 0; i < 2*width + 1; ++i) {
		cout << "-";
	}
	cout << endl;
	for (int i = 0; i < height; ++i) {
		if (i < 10) {
				cout << " ";
			}
		cout << i << "|" << " ";
		for (int j = 0; j < width; ++j) {
			cout << mask[i][j] << " ";
		}
		cout << endl;
	}
	cout << endl;
}
//For debugging
void Board::PrintBoard() {
	for (int i = 0; i < this->height; ++i) {
		cout << endl;
		for (int j = 0; j < this->width; ++j) {
			cout << this->board[i][j] << " ";
		}
	}
	cout << endl;
}

File: Golang/Minesweeper/board/board.go
package board

import(
	"fmt"
	"os"
	"bufio"
	"strings"
	"strconv"
	"math/rand"
	"time"
)

type Board struct {
	Width int
	Height int
	MineCount int
	Values [][]int
	Mask [][]string
}

func Initialize(playBoard *Board) {
	GetInput(playBoard)

	playBoard.Values = make([][]int, playBoard.Width)
    for i := range playBoard.Values {
        playBoard.Values[i] = make([]int, playBoard.Height)
    }

    playBoard.Mask = make([][]string, playBoard.Width)
    for i := range playBoard.Mask {
        playBoard.Mask[i] = make([]string, playBoard.Height)
    }

    for i := 0; i < playBoard.Height; i++ {
		for j := 0; j < playBoard.Width; j++ {
			playBoard.Mask[j][i] = "#"
		}
	}

    PlaceMines(playBoard)
}

func GetInput(playBoard *Board) {
	var errW error
	var errH error
	var errM error

	fmt.Println("Chose Width, Height and Number of mines separated by space:")
	reader := bufio.NewReader(os.Stdin)

	for{
		input, _ := reader.ReadString('\n')
		input = strings.TrimRight(input, " \t\r\n")
		inputs := strings.Split(input, " ")

		if len(inputs) != 3 {
			fmt.Println("Invalid input. Try again.")
			continue
		}

		playBoard.Width, errW = strconv.Atoi(inputs[0])
		playBoard.Height, errH = strconv.Atoi(inputs[1])
		playBoard.MineCount, errM = strconv.Atoi(inputs[2])

		if errW != nil || errH != nil || errM != nil {
			fmt.Println("Invalid input. Try again.")
			continue
		}
		break
	}
}

func PlaceMines(playBoard *Board) {
	rand.Seed( time.Now().UTC().UnixNano())
	minesPlaced := 0

	for minesPlaced < playBoard.MineCount {
		randW := rand.Int() % playBoard.Width
		randH := rand.Int() % playBoard.Height

		if playBoard.Values[randH][randW] == 0 {
			playBoard.Values[randH][randW] = 9
			minesPlaced++
		}
	}
}

func PrintMask(playBoard Board) {
	fmt.Printf("    ")
	for i := 0; i < playBoard.Width; i++ {
		fmt.Printf("%d", i%10)
		fmt.Printf(" ")
	}
	fmt.Printf("\n  ")
	for i := 0; i < 2*playBoard.Width + 1; i++ {
		fmt.Printf("-")
	}
	for i := 0; i < playBoard.Height; i++ {
		fmt.Printf("\n")
		if i < 10 {
			fmt.Printf(" ")
		}
		fmt.Printf("%d", i)
		fmt.Printf("| ")
		for j := 0; j < playBoard.Width; j++ {
			fmt.Printf(playBoard.Mask[j][i])
			fmt.Printf(" ")	
		}
	}
}
// For debugging
func PrintBoard(playBoard Board) {
	for i := 0; i < playBoard.Height; i++ {
		fmt.Printf("\n")
		for j := 0; j < playBoard.Width; j++ {
			fmt.Printf("%d", playBoard.Values[j][i])
			fmt.Printf(" ")	
		}
	}
}